
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>CUT &amp; RUN pipeline &#8212; HemTools  documentation</title>
    <link rel="stylesheet" href="../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Single-end ChIP-seq" href="chip_seq_single.html" />
    <link rel="prev" title="ATAC-seq" href="atac_seq.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chip_seq_single.html" title="Single-end ChIP-seq"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="atac_seq.html" title="ATAC-seq"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">HemTools  documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="NGS_pipelines.html" accesskey="U">A collection of NGS_pipelines</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="cut-run-pipeline">
<h1>CUT &amp; RUN pipeline<a class="headerlink" href="#cut-run-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">HemTools</span> <span class="n">cut_run</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">j</span> <span class="n">JID</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">short</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="n">INPUT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="n">DESIGN_MATRIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">guess_input</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INDEX_FILE</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">-</span><span class="n">g</span> <span class="n">GENOME</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BLACKLIST</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">CHROM_SIZE</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EFFECTIVEGENOMESIZE</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h2>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-j, --jid</kbd></td>
<td><p class="first">enter a job ID, which is used to make a new directory. Every output will be moved into this folder.</p>
<p class="last">Default: “{{subcmd}}_yli11_2019-04-14”</p>
</td></tr>
<tr><td class="option-group">
<kbd>--short</kbd></td>
<td><p class="first">Force to use the short queue. (only if R1+R2 fastq.gz size &lt;=250M)</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--debug</kbd></td>
<td><p class="first">Not for end-user.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>-f, --input</kbd></td>
<td>tab delimited 3 columns (tsv file): Read 1 fastq, Read 2 fastq, sample ID</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-d, --design_matrix</kbd></td>
</tr>
<tr><td>&#160;</td><td>tab delimited 3 columns (tsv file): treatment sample ID, control sample ID, peakcall ID</td></tr>
<tr><td class="option-group">
<kbd>--guess_input</kbd></td>
<td><p class="first">Let the program generate the input files for you.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Genome Info">
<h2>Genome Info<a class="headerlink" href="#Genome Info" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd>-i, --index_file</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">BWA index file</p>
<p class="last">Default: “D:HemToolssubcmd/../hg19/bwa_16a_index/hg19.fa”</p>
</td></tr>
<tr><td class="option-group">
<kbd>-g, --genome</kbd></td>
<td><p class="first">genome version: hg19, hg38, mm10, mm9.</p>
<p class="last">Default: “hg19”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-b, --Blacklist</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Blacklist file</p>
<p class="last">Default: “D:HemToolssubcmd/../hg19/Hg19_Blacklist.bed”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-s, --chrom_size</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">chrome size</p>
<p class="last">Default: “D:HemToolssubcmd/../hg19/hg19.chrom.sizes”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-e, --effectiveGenomeSize</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">effectiveGenomeSize for bamCoverage</p>
<p class="last">Default: “2451960000”</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Go to your data directory and type the following.</p>
<p><strong>Step 0: Load python version 2.7.13.</strong></p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ module load python/2.7.13
</pre></div>
</div>
<p><strong>Step 1: Prepare input files, generate fastq.tsv.</strong></p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ HemTools cut_run --guess_input

    Input fastq files preparation complete! ALL GOOD!
    Please check if you like the computer-generated labels in : fastq.tsv
    Input peakcall file preparation complete! File name: peakcall.tsv
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you are preparing fastq.tsv and peakcall.tsv yourself, please make sure <code class="docutils literal notranslate"><span class="pre">no</span> <span class="pre">space</span> <span class="pre">anywhere</span></code> in the file. Note that the seperator is tab. Spaces in file name will cause errors.</p>
</div>
<p><strong>Step 2: Check the computer-generated input list (manually), make sure they are correct.</strong></p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ less fastq.tsv

$ less peakcall.tsv
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">a random string will be added to the generated files (e.g., fastq.94c049cbff1f.tsv) if they exist before running step 1.</p>
</div>
<p><strong>Step 3: Submit your job.</strong></p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ HemTools cut_run -f fastq.tsv -d peakcall.tsv
</pre></div>
</div>
</div>
<div class="section" id="report-bug">
<h2>Report bug<a class="headerlink" href="#report-bug" title="Permalink to this headline">¶</a></h2>
<p>Once the job is finished, you will be notified by email with some attachments.  If no attachment can be found, it might be caused by an error. In such case, please go to the result directory (where the log_files folder is located) and type:</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ HemTools report_bug
</pre></div>
</div>
</div>
<div class="section" id="use-different-genome-index">
<h2>Use different genome index<a class="headerlink" href="#use-different-genome-index" title="Permalink to this headline">¶</a></h2>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ HemTools cut_run -f fastq.tsv -d peakcall.tsv -i YOUR_GENOME_INDEX
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference external" href="https://github.com/YichaoOU/HemTools/">code &#64; github</a>.</td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">CUT &amp; RUN pipeline</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
<li><a class="reference internal" href="#Genome Info">Genome Info</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#report-bug">Report bug</a></li>
<li><a class="reference internal" href="#use-different-genome-index">Use different genome index</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="atac_seq.html"
                        title="previous chapter">ATAC-seq</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="chip_seq_single.html"
                        title="next chapter">Single-end ChIP-seq</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../_sources/content/NGS_pipelines/cut_run.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chip_seq_single.html" title="Single-end ChIP-seq"
             >next</a> |</li>
        <li class="right" >
          <a href="atac_seq.html" title="ATAC-seq"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">HemTools  documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="NGS_pipelines.html" >A collection of NGS_pipelines</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Yichao Li, Yong Cheng.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>