Footprint analysis for ATAC-seq data
===================================

::

	usage: atac_seq_footprint.py [-h] [-j JID] [--pipeline_type PIPELINE_TYPE] -f
	                             INPUT [-g GENOME]

	RGT_HINT atac-seq footprint with bias correction

	optional arguments:
	  -h, --help            show this help message and exit
	  -j JID, --jid JID     enter a job ID, which is used to make a new directory.
	                        Every output will be moved into this folder. (default:
	                        atac_seq_footprint_yli11_2019-08-05)
	  --pipeline_type PIPELINE_TYPE
	                        Not for end-user. (default: atac_seq_footprint)
	  -f INPUT, --input INPUT
	                        3-col tsv, bam,bed,output-prefix (default: None)

	Genome Info:
	  -g GENOME, --genome GENOME
	                        genome version: hg19, hg38, mm9, mm10. By default,
	                        specifying a genome version will automatically update
	                        index file, black list, chrom size and
	                        effectiveGenomeSize, unless a user explicitly sets
	                        those options. (default: hg19)


Summary
^^^^^^^

Currently, this pipeline applies HINT-ATAC and output bias-corrected footprint bed files and cutsites bw files.

As of 8-5-2019, motif pwm is not included.

Input
^^^^^

The input file is a tsv format containing 3 columns: bam, bed, output-prefix.

Either relative path or absolute path is OK.

Suppose you run this pipeline in ``bam_files`` folder generated by ``HemTools atac_seq``

::

	Hudep1.markdup.bam	../peak_files/Hudep1.markdup.rmchrM_peaks.narrowPeak	H1
	Hudep2.markdup.bam	../peak_files/Hudep2.markdup.rmchrM_peaks.narrowPeak	H2


Usage
^^^^^

.. code:: bash

	module load python/2.7.13

	atac_seq_footprint.py -f input.list


Reference
^^^^^^^^^

https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1642-2

























