Search.setIndex({docnames:["content/Differential_analysis/diffpeaks","content/Installation/Installation","content/Motif_analysis/methylmotifs","content/NGS_pipelines/NGS_pipelines","content/NGS_pipelines/atac_seq","content/NGS_pipelines/chip_seq_pair","content/NGS_pipelines/chip_seq_single","content/NGS_pipelines/crispr_seq","content/NGS_pipelines/cut_run","index"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.cpp":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,"sphinx.ext.viewcode":1,sphinx:54},filenames:["content\\Differential_analysis\\diffpeaks.rst","content\\Installation\\Installation.rst","content\\Motif_analysis\\methylmotifs.rst","content\\NGS_pipelines\\NGS_pipelines.rst","content\\NGS_pipelines\\atac_seq.rst","content\\NGS_pipelines\\chip_seq_pair.rst","content\\NGS_pipelines\\chip_seq_single.rst","content\\NGS_pipelines\\crispr_seq.rst","content\\NGS_pipelines\\cut_run.rst","index.rst"],objects:{},objnames:{},objtypes:{},terms:{"250m":4,"94c049cbff1f":3,"case":3,"default":4,"new":4,"short":4,Not:4,_yli11_2019:4,adapt:4,adaptor_i:4,adaptor_x:4,added:3,all:[3,9],alwai:[3,9],api:1,atac:[3,9],atac_seq:[3,4],attach:3,avail:[3,9],bamcoverag:4,bed:4,befor:3,blacklist:4,bug:9,bwa:4,bwa_16a_index:4,can:[3,9],caus:3,check:3,chip:[3,9],chrom:4,chrom_siz:4,chrome:4,code:[0,1,2,3,4,5,6,7,8,9],column:4,command:[3,9],complet:3,comput:3,control:3,correct:3,ctgtctcttatacacatct:4,cut:[3,9],cut_run:9,data:3,debug:4,delimit:4,directori:[3,4],discoveri:2,effectivegenomes:4,email:3,end:[3,4,9],enter:4,error:3,everi:4,exist:3,fals:4,fastq:[3,4,9],file:[3,4],finish:3,folder:[3,4],follow:3,forc:4,found:3,free:1,gener:[3,4],github:[0,1,2,3,4,5,6,7,8,9],good:3,guess_input:[3,4],hemtool:3,hemtoolssubcmd:4,hg19:4,hg19_blacklist:4,hg38:4,index:4,index_fil:4,input:[3,4],instal:9,jid:4,job:[3,4],label:3,less:3,let:4,like:[3,9],list:3,load:[3,9],locat:3,log_fil:3,look:9,make:[3,4],manual:3,methyl:2,might:3,mm10:4,mm9:4,modul:[3,9],most:1,motif:2,move:4,name:3,need:3,ngs_pipelin:9,notifi:3,onc:3,onli:4,output:4,pair:[3,9],peakcal:[3,9],pipelin:3,pleas:3,prepar:3,program:4,python:[3,9],queue:4,random:3,read:4,report:9,report_bug:3,result:3,run:[3,9],sampl:4,see:[3,9],seq:[3,9],sequenc:4,sinc:3,singl:[3,9],size:4,some:3,step:3,string:3,sub:[3,9],subcmd:[3,4],submit:3,sure:3,tab:4,thei:3,thi:[4,9],tsv:[3,4,9],type:3,typic:9,usag:[4,9],use:4,used:4,user:4,version:[3,4],where:3,which:4,you:[3,4,9],your:3},titles:["Differential Peaks pipeline","Installation","&lt;no title&gt;","A collection of NGS_pipelines","ATAC-seq","Paired-end ChIP-seq","Single-end ChIP-seq","CRISPR-seq","CUT &amp; RUN pipeline","HemTools: <em>a collection of NGS pipelines and bioinformatic analyses</em>"],titleterms:{NGS:9,adaptor:4,analys:9,argument:4,atac:4,bioinformat:9,bug:3,chip:[5,6],collect:[3,9],crispr:7,cut:8,differenti:0,end:[5,6],gener:9,genom:4,hemtool:9,info:4,instal:1,name:4,ngs_pipelin:3,pair:5,peak:0,pipelin:[0,8,9],principl:9,report:3,run:8,seq:[4,5,6,7],singl:6,typic:3,usag:3}})